File: examples/leetcode-out/cpp/28/find-the-index-of-the-first-occurrence-in-a-string.cpp
#include <bits/stdc++.h>
using namespace std;

string _indexString(const string& s, int i) {
	int n = s.size();
	if (i < 0) i += n;
	if (i < 0 || i >= n) throw std::out_of_range("index out of range");
	return string(1, s[i]);
}

int strStr(string haystack, string needle){
	auto n = haystack.size();
	auto m = needle.size();
	if (m == 0) {
		return 0;
	}
	if (m > n) {
		return -1;
	}
	for (int i = 0; i < n - m + 1; i++) {
		auto j = 0;
		while (j < m) {
			if (_indexString(haystack, i + j) != _indexString(needle, j)) {
				break;
			}
			j = j + 1;
		}
		if (j == m) {
			return i;
		}
	}
	return -1;
}

int main() {
	return 0;
}

File: examples/leetcode-out/dart/28/find-the-index-of-the-first-occurrence-in-a-string.dart
int strStr(String haystack, String needle) {
	dynamic n = haystack.length;
	dynamic m = needle.length;
	if ((m == 0)) {
		return 0;
	}
	if ((m > n)) {
		return -1;
	}
	for (var i = 0; i < ((n - m) + 1); i++) {
		dynamic j = 0;
		while ((j < m)) {
			if ((_indexString(haystack, ((i + j)).toInt()) != _indexString(needle, j))) {
				break;
			}
			j = ((j + 1)).toInt();
		}
		if ((j == m)) {
			return i;
		}
	}
	return -1;
}

void main() {
}

String _indexString(String s, int i) {
	var runes = s.runes.toList();
	if (i < 0) {
		i += runes.length;
	}
	if (i < 0 || i >= runes.length) {
		throw RangeError('index out of range');
	}
	return String.fromCharCode(runes[i]);
}

